docs: update versioning and deprecation policies (#51423)
angular/angular: aio/content/guide/releases.md

We strive to ensure that future changes are always introduced in a predictable way.
We want everyone who depends on Angular to know when and how new features are added, and to be well-prepared when obsolete ones are removed.

> Sometimes "breaking changes", such as the removal of support for select APIs and features, are necessary to innovate and stay current with new best practices, changing dependencies, or changes in the \(web\) platform itself. Each such "breaking change" goes through a deprecation process which we've explained below in our [deprecation policy](#deprecation-policy).

Reviewer: ```suggestion
Sometimes *breaking changes*, such as the removal of APIs or features, are necessary to innovate and stay current with evolving best practices, changing dependencies, or shifts in the web platform. These breaking changes go through a deprecation process explained in our [deprecation policy](#deprecation-policy).
```

some suggestions here mostly for brevity and some word choice

> To make these transitions as straightforward as possible, we make these commitments to you:

Reviewer: ```suggestion
To make these transitions as straightforward as possible, the Angular team makes these commitments:
```
nit: we want to avoid first person language ("we") in technical docs as much as possible

<a id="deprecation"></a>

> ## Deprecation practices

Reviewer 2: Fyi, by changing this section name you broke several links (look for `guide/releases#deprecation-practices`).

> We consider APIs for deprecation if they are obsolete, discouraging continued or future use, either because there is a preferred alternative or because there is an intention remove them in the future. The APIs are still available through the deprecated phase. Deprecation is usually performed well in advance of their actual removal.

Reviewer: ```suggestion
When the Angular team intends to remove an API or feature, it will be marked as *deprecated*. This occurs when when an API is obsolete, superseded by another API, or otherwise discontinued. Deprecated API remain available through their deprecated phase, which lasts a minimum two major versions (approximately one year).
```

My thinking is that we should go with a definition of deprecated that is explicitly about intent to remove (rather than just to mark something as discouraged).

To help ensure that you have sufficient time and a clear path to update, this is our deprecation policy:

| Deprecation stages | Details |
|:---                |:---     |
| Announcement       | We announce deprecated APIs and features in the [change log](https://github.com/angular/angular/blob/main/CHANGELOG.md "Angular change log"). Deprecated APIs appear in the [documentation](api?status=deprecated) with ~~strikethrough~~. When we announce a deprecation, we also announce a recommended update path. For convenience, [Deprecations](guide/deprecations) contains a summary of deprecated APIs, and features. Additionally, all deprecated APIs are annotated with `@deprecated` in the corresponding documentation, which enables text editors and IDEs to provide hints if your project depends on them.                            |

> | Deprecation period | When an API or a feature is deprecated, it is still present in at least the next two major releases (period of at least 12 months). After that, deprecated APIs and features are candidates for removal. A deprecation can be announced in any release, but the removal of a deprecated API or feature happens only in major release. Until a deprecated API or feature is removed, it is maintained according to the LTS support policy, meaning that only critical and security issues are fixed. |

Reviewer: Can we get away with saying "approximately one year"? We've had situations before where schedule variations led to a the time between two major versions being just slightly less than one full year.

Author: I'd love to set the expectation it's 12+ months. Approximately one year could be also interpreted as 11 months and 25 days, or 11 months.

| npm dependencies   | We only make npm dependency updates that require changes to your applications in a major release. In minor releases, we update peer dependencies by expanding the supported versions, but we do not require projects to update these dependencies until a future major version. This means that during minor Angular releases, npm dependency updates within Angular applications and libraries are optional.                                               |

<a id="public-api"></a>

## Compatibility policy

Angular is a collection of many packages, subprojects, and tools.
To prevent accidental use of private APIs and so that you can clearly understand what is covered by the practices described here &mdash; we document what is and is not considered our public API surface.
For details, see [Supported Public API Surface of Angular](https://github.com/angular/angular/blob/main/docs/PUBLIC_API.md "Supported Public API Surface of Angular").

To guarantee backward compatibility of Angular we run a series of checks before we merge any change:

* Unit tests and integration tests
* Comparing the type definitions of the public API surface before and after the change

> * Running the tests of all the applications at Google that depend on Angular

Reviewer: > Running the tests of all the applications at Google that depend on Angular

This isn't exactly true; we don't run ALL the tests on every change for framework. Maybe something like "A large suite of tests for Google applications built with Angular"

Author: Yeah, on every PR we run subset of the tests, but ultimately we'll learn if any of the tests fails. We still get validation with all the g3, that's why I phrased it this way.

> Any changes to the public API surface are done using the versioning, support, and depreciation policies previously described. In the case of an exception where a critical fix part of a major or a patch release introduces a backward incompatible change to the public API surface, we'll publish explicit communication.
> 

Reviewer: ```suggestion
Any changes to the public API surface are made in accordance with the versioning, support, and depreciation policies previously described. In exceptional cases, such as critical security patches, fixes may introduce backwards incompatible changes. Such exceptional cases are accompanied by explicit notice on the framework's official communication channels.
```

Do we need to specify what those official communication channels are?

<a id="updating"></a>

## Breaking change policy and update paths

> Breaking change requires you to do work because the state after it is not backward compatible with the state before it. You can find the rare exceptions from this rule in the [Compatibility policy](#compatibility-policy). Examples of breaking changes are the removal of public APIs or other changes of the type definition of Angular, changing the timing of calls, or updating to a new version of a dependency of Angular, which includes breaking changes itself.

Reviewer: "requires you to do work" feels like it needs more nuance- we often make breaking changes that don't require manual intervention if you use `ng update`.

Author: Still requires you to run `ng update`, which is a manual intervention.

<a id="deprecation"></a>

## Deprecation policy

When the Angular team intends to remove an API or feature, it will be marked as *deprecated*. This occurs when when an API is obsolete, superseded by another API, or otherwise discontinued. Deprecated API remain available through their deprecated phase, which lasts a minimum two major versions (approximately one year).

To help ensure that you have sufficient time and a clear path to update, this is our deprecation policy:

| Deprecation stages | Details |
|:---                |:---     |
| Announcement       | We announce deprecated APIs and features in the [change log](https://github.com/angular/angular/blob/main/CHANGELOG.md "Angular change log"). Deprecated APIs appear in the [documentation](api?status=deprecated) with ~~strikethrough~~. When we announce a deprecation, we also announce a recommended update path. For convenience, [Deprecations](guide/deprecations) contains a summary of deprecated APIs, and features. Additionally, all deprecated APIs are annotated with `@deprecated` in the corresponding documentation, which enables text editors and IDEs to provide hints if your project depends on them.                            |

> | Deprecation period | When an API or a feature is deprecated, it is still present in at least the next two major releases (period of at least 12 months). After that, deprecated APIs and features are candidates for removal. A deprecation can be announced in any release, but the removal of a deprecated API or feature happens only in major release. Until a deprecated API or feature is removed, it is maintained according to the LTS support policy, meaning that only critical and security issues are fixed. |

Reviewer: ```suggestion
| Deprecation period | When an API or a feature is deprecated, it will remain in a supported release for a period of at least 12 months. A deprecation can be announced in any release, but the removal of a deprecated API or feature happens only in major release. Until a deprecated API or feature is removed, it is maintained according to the LTS support policy, meaning that only critical and security issues are fixed. |
```
being pedantic: the policy is 12 months of _support_, which is distinct from still appearing in two major releases (which, if we're saying 12 months, isn't _strictly_ the case, but will be true most of the time)
